sdk/core/src/error/mod.rs

```diff
@@ -0,0 +1,371 @@
+use std::borrow::Cow;
+use std::fmt::{Debug, Display};
+
+mod azure_core_errors;
+mod hyperium_http;
+mod macros;
+
+/// A convience alias for `Result` where the error type is hard coded to `Error`
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// The kind of error
+///
+/// The classification of error is intentionally fairly coarse.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum ErrorKind {
+    /// An HTTP status code that was not expected
+    HttpResponse { status: u16 },
+    /// An error performing IO
+    Io,
+    /// An error converting data
+    DataConversion,
+    /// An error getting an API credential token
+    Credential,
+    #[cfg(feature = "mock_transport_framework")]
+    /// An error having to do with the mock framework
+    MockFramework,
```

Comment:
Is there any reason this should just be an `HttpResponse`. That's what we do in other libraries. The goal of the mock HTTP frameworks across SDKs is to be transparent. So why should any match expression care whether a failed request is from the real HTTP transport or a mock?

Reply:
Errors in the mock framework can happen that have nothing to do with an HttpResponse. For example, failing to read some of the mock data or failing to provide mock data. This is for those sort of errors.

Reply:
So a mocked service failure will still return a FailedResponse?

Reply:
Yes